crates/ruff_server/src/lint.rs

```diff
@@ -108,25 +106,51 @@ pub(crate) fn check(
     );
 
     let noqa_edits = generate_noqa_edits(
-        &document_path,
-        diagnostics.as_slice(),
+        document_path,
+        data.as_slice(),
         &locator,
         indexer.comment_ranges(),
         &linter_settings.external,
         &directives.noqa_line_for,
         stylist.line_ending(),
     );
 
-    diagnostics
+    let mut diagnostics = Diagnostics::default();
+
+    // Populate all cell URLs with an empty diagnostic list.
+    // This ensures that cells without diagnostics still get updated.
+    if let Some(notebook) = query.as_notebook() {
+        for url in notebook.urls() {
+            diagnostics.entry(url.clone()).or_default();
+        }
+    }
+
+    for (index, diagnostic) in data
         .into_iter()
         .zip(noqa_edits)
-        .map(|(diagnostic, noqa_edit)| to_lsp_diagnostic(diagnostic, noqa_edit, document, encoding))
-        .collect()
+        .map(|(diagnostic, noqa_edit)| {
+            to_lsp_diagnostic(diagnostic, &noqa_edit, &source_kind, &index, encoding)
+        })
+    {
+        if let Some(notebook) = query.as_notebook() {
+            let Some(uri) = notebook.cell_uri_by_index(index) else {
+                tracing::warn!("Unable to find notebook cell at index {index}.");
+                continue;
+            };
+            diagnostics.entry(uri.clone()).or_default().push(diagnostic);
+        } else {
+            diagnostics
+                .entry(query.make_key().into_url())
+                .or_default()
+                .push(diagnostic);
+        }
+    }
+
+    diagnostics
```

Comment:
nit: would it be possible to combine this in a single `if` condition? It would be easier to reason about when everything related to notebook is in the same block and has access to the notebook.

```rs
if let Some(notebook) = query.as_notebook() {
	// .. do everything with the notebook
} else {
	// .. do everything with the text document
}
```

You can possible store the LSP diagnostics in a variable:
```rs
    let mut lsp_diagnostics = data
        .into_iter()
        .zip(noqa_edits)
        .map(|(diagnostic, noqa_edit)| {
            to_lsp_diagnostic(diagnostic, &noqa_edit, &source_kind, &index, encoding)
        });
```